### PR TITLE
connect error, epoll_ctl failed: file not exist.

### PR DESCRIPTION
**Pull request by zlianzhuang**
because we use async connect module. when pgStatus is not
CONNECTION_OK/CONNECTION_BAD we need get the envent from
pollMode(PQconnectPoll), but we init it to
'WL_SOCKET_READABLE | WL_SOCKET_WRITEABLE' in ManageWorkerPool,
and the PQSocket maybe close so we shoud rebuild the event.

### CODE REVIEW

**Comment by zlianzhuang** [comment]
hi. why the ci is still Required?

**Comment by JelteF** [comment]
Thanks a lot for the PR. I fixed our CI settings to build pull requests from forks as well. It seems you haven't run `citus_indent` before pushing, you can install it through here: https://github.com/citusdata/tools/tree/develop/uncrustify

**Comment by serprex** [comment]
@JelteF fixed formatting

File: src/backend/distributed/executor/adaptive_executor.c [hunk 1]
```diff
@@ -2121,8 +2123,30 @@ ManageWorkerPool(WorkerPool *workerPool)
 		/* create a session for the connection */
 		session = FindOrCreateWorkerSession(workerPool, connection);
 
+		pollMode = PQconnectPoll(connection->pgConn);
```
Comment (marcocitus) [inline src/backend/distributed/executor/adaptive_executor.c]: What if we call ConnectionStateMachine here?

Comment (zlianzhuang) [reply src/backend/distributed/executor/adaptive_executor.c]: yes. this is right.  call ConnectionStateMachine(session) is well.

Comment (zlianzhuang) [reply src/backend/distributed/executor/adaptive_executor.c]: i have rewrite the code used ConnectionStateMachine.

File: src/backend/distributed/executor/adaptive_executor.c [hunk 2]
```diff
@@ -2373,6 +2371,11 @@ ConnectionStateMachine(WorkerSession *session)
 				}
 
 				pollMode = PQconnectPoll(connection->pgConn);
+
+				if (status == CONNECTION_CHECK_WRITABLE && pollMode != PGRES_POLLING_OK)
+				{
+					execution->connectionSetChanged = true; /* can't break, we need the changed waitflags */
```
Comment (marcocitus) [inline src/backend/distributed/executor/adaptive_executor.c]: CONNECTION_CHECK_WRITABLE primarily relates to target_session_attrs=read-write, which I don't think we really support right now. Do we need this check?

Comment (zlianzhuang) [reply src/backend/distributed/executor/adaptive_executor.c]: we can set environment PGTARGETSESSIONATTRS or modify func NodeConninfoGucCheckHook for accept target_session_attrs to allow the cn connect the writeable worker when then nodename(pg_dist_node) have two ips。
if the first ip is read only. libpq server will close the connect and connect the next ip. libpq client receive status  CONNECTION_CHECK_WRITABLE pollMode PGRES_POLLING_WRITING and wait for the response(not receive any failed). but BuildWaitEventSet will use the PQsocket and it is closed the new socket is the next ip socket. so we should rebuild the event to get the new socket. ModifyWaitEvent is still operate the old socket what cause a epoll_ctl error.

File: src/backend/distributed/connection/connection_management.c [hunk 3]
```diff
@@ -528,26 +535,34 @@ MultiConnectionStatePoll(MultiConnectionPollState *connectionState)
 
 	connectionState->pollmode = PQconnectPoll(connection->pgConn);
 
+	if (status == CONNECTION_CHECK_WRITABLE && connectionState->pollmode !=
```
[the remark this answers is not in the document]
Comment (zlianzhuang) [reply src/backend/distributed/connection/connection_management.c]: sorry, I haven't seen my email lately, i will do it later.

File: src/backend/distributed/connection/connection_management.c [hunk 4]
```diff
@@ -59,6 +59,12 @@ enum MultiConnectionPhase
 	MULTI_CONNECTION_PHASE_CONNECTED,
 	MULTI_CONNECTION_PHASE_ERROR,
 };
+typedef enum MultiConnectionStateChanged
+{
+	MULTI_CONNECTION_STATE_CHANGED_NO_CHANGED,
```
[the remark this answers is not in the document]
Comment (zlianzhuang) [reply src/backend/distributed/connection/connection_management.c]: done. and push the code.

File: src/backend/distributed/connection/connection_management.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (zlianzhuang) [reply src/backend/distributed/connection/connection_management.c]:
```
if (multiConnection->socketInTheLastPoll != INVALID_SOCKET && multiConnection->socketInTheLastPoll != PQSocket(conn))
```

we maybe get the same socket. `"socketInTheLastPoll != PQSocket(conn))`  don't know the socket is changed.

[the remark this answers is not in the document]
Comment (zlianzhuang) [reply src/backend/distributed/connection/connection_management.c]: the socket fd is 9, the new socke is also maybe 9.

File: src/backend/distributed/executor/adaptive_executor.c [hunk 5]
```diff
@@ -2120,9 +2120,7 @@ ManageWorkerPool(WorkerPool *workerPool)
 
 		/* create a session for the connection */
 		session = FindOrCreateWorkerSession(workerPool, connection);
-
-		/* always poll the connection in the first round */
-		UpdateConnectionWaitFlags(session, WL_SOCKET_READABLE | WL_SOCKET_WRITEABLE);
+		ConnectionStateMachine(session);
```
Comment (marcocitus) [inline src/backend/distributed/executor/adaptive_executor.c]: I merged this part of the PR.

Comment (zlianzhuang) [reply src/backend/distributed/executor/adaptive_executor.c]: yes，i use the pg_auto_failover. adaptive/real-time/task-tracker test is success（select * from aa ; select * from aa where id = xx），

**Comment by onderkalaci** [comment]
@zlianzhuang Is it possible for you to test this branch and see if it fixes the issue you see: https://github.com/citusdata/citus/pull/3811

Thanks

**Comment by onderkalaci** [comment]
@zlianzhuang The fix is on our newest version Citus 9.3, so you may test with that..

**Pull request closed without merging**